verilog/tools/ls/autoexpand.cc

```diff
@@ -243,27 +362,265 @@ void AutoExpander::Module::EmitPortConnections(
+    if (net_name.end() <= auto_span.begin() &&
+        net_name.begin() >= auto_span.end()) {
+      declared_wires.insert(net_name);
+    }
+  }
+  const auto is_undeclared_connected_and_not_wire = [&](const Port &port) {
+    return port.declaration == Port::Undeclared && port.connection &&
+           !declared_wires.contains(port.name);
+  };
+
   for (const Port &port : ports_) {
-    if (!pred(port)) continue;
-    if (first) {
-      if (output.tellp() != 0) output << ',';
-      output << '\n' << indent << "// " << header;
-      first = false;
-    } else {
-      output << ',';
+    if (port.direction == Port::Inout &&
+        is_undeclared_connected_and_not_wire(port)) {
+      output << indent << "wire " << port.name << ";  // To/From "
+             << port.connection->instance << " of " << port.connection->type
+             << "\n";
+    }
+  }
+  for (const Port &port : ports_) {
+    if (port.direction == Port::Output &&
+        is_undeclared_connected_and_not_wire(port)) {
+      output << indent << "wire " << port.name << ";  // From "
+             << port.connection->instance << " of " << port.connection->type
+             << "\n";
     }
-    auto connected_name = port.name;
+  }
+}
+
+void AutoExpander::Module::EmitUnconnectedOutputRegDeclarations(
+    std::ostream &output, const absl::string_view indent,
+    const absl::string_view auto_span) const {
+  absl::flat_hash_set<absl::string_view> declared_regs;
+  for (const auto &reg : FindAllRegisterVariables(symbol)) {
+    const SyntaxTreeLeaf *const reg_name_leaf =
+        GetNameLeafOfRegisterVariable(*reg.match);
+    const absl::string_view reg_name = reg_name_leaf->get().text();
+    if (reg_name.end() <= auto_span.begin() &&
+        reg_name.begin() >= auto_span.end()) {
+      declared_regs.insert(reg_name);
+    }
+  }
+  const auto is_declared_unconnected_and_not_reg = [&](const Port &port) {
+    return port.declaration == Port::Declared && !port.connection &&
+           !declared_regs.contains(port.name);
+  };
+
+  for (const Port &port : ports_) {
+    if (port.direction == Port::Output &&
+        is_declared_unconnected_and_not_reg(port)) {
+      output << indent << "reg " << port.name << ";\n";
+    }
+  }
+}
+
+void AutoExpander::Module::GenerateConnections(
+    const Template *tmpl,
+    const std::function<void(const Port &, absl::string_view)> &fun) const {
+  for (const Port &port : ports_) {
+    absl::string_view connected_name = port.name;
     if (tmpl) {
-      auto it = tmpl->connections.find(port.name);
+      const auto it = tmpl->connections.find(port.name);
       if (it != tmpl->connections.end()) connected_name = it->second;
     }
-    output << '\n'
-           << indent << '.' << port.name << '(' << connected_name << ")";
+    fun(port, connected_name);
+  }
+}
+
+void AutoExpander::Module::AddGeneratedConnection(
+    const absl::string_view port_name, const Port::Direction direction,
+    const Connection connection) {
```

Comment:
Connection is a fairly small struct, but I'd probably still pass it by `const &` not by value (mostly to not slow down readability flow - if it is passed by value the reader might stop a second and wonder if there is a particular reason).